### PR TITLE
Change/refactor needs

### PR DESCRIPTION
**Pull request by logxen**
Major refactor that also adds PowerSpawns

### CODE REVIEW

**Comment by karlthepagan** [comment]
> refactored newTarget into a priority array

I'm excited. ;)

**Comment by logxen** [comment]
well ... lol I failed at the syntax there and went to an if-else list in the next commit but it's setup so we can do that easily as soon as I figure out how to declare it all correctly. :)

**Comment by chobobobo** [comment]
It looks like my creeps are trying to reallocate resources in a room that only has a storage but no terminal.  The little <=> keeps appearing above their heads.

**Comment by logxen** [comment]
@karlthepagan ok latest push should fix that. I need to do tests without internal. :P

**Comment by karlthepagan** [comment]
```
TypeError: Cannot read property 'find' of undefined
    at StructurePowerSpawn.getNeeds (extensions:302:66)
    at Action.action.newTargetPowerSpawn (creep.action.reallocating:194:33)
    at Action.action.newTarget (creep.action.reallocating:393:48)
    at Action.assign (creep.Action:89:50)
    at Object.mod.nextAction (creep.behaviour.hauler:55:20)
    at Object.mod.run (creep.behaviour.hauler:8:14)
    at Creep.run (creep:47:39)
    at run (creep:447:30)
```

**Comment by Dasku** [comment]
Testing this now it seems that nuker and powerspawn are being saved as memory.nuker/powerSpawn, however in setStore it is looking for containers in memory.resources.nuker/powerSpawn

When I try and set orders on the powrerspawn it gives -7

File: global.js [hunk 1]
```diff
@@ -139,45 +143,12 @@ mod.LAB_MASTER = 'master';
...
-    [RESOURCE_CATALYZED_LEMERGIUM_ALKALIDE]: [RESOURCE_LEMERGIUM_ALKALIDE, RESOURCE_CATALYST],
-    [RESOURCE_CATALYZED_ZYNTHIUM_ACID]: [RESOURCE_ZYNTHIUM_ACID, RESOURCE_CATALYST],
-    [RESOURCE_CATALYZED_ZYNTHIUM_ALKALIDE]: [RESOURCE_ZYNTHIUM_ALKALIDE, RESOURCE_CATALYST],
-    [RESOURCE_CATALYZED_GHODIUM_ACID]: [RESOURCE_GHODIUM_ACID, RESOURCE_CATALYST],
-    [RESOURCE_CATALYZED_GHODIUM_ALKALIDE]: [RESOURCE_GHODIUM_ALKALIDE, RESOURCE_CATALYST],
-};
+mod.LAB_REACTIONS = {};
+for(var a in REACTIONS){
+    for(var b in REACTIONS[a]){
+        mod.LAB_RECIPES[REACTIONS[a][b]] = [a,b];
```
Comment (Spedwards) [inline global.js]: Should be `mod.LAB_REACTIONS[REACTIONS[a][b]] = [a, b];`

Also, it should be `let a` and `let b`.

File: creep.action.reallocating.js [hunk 2]
```diff
@@ -190,7 +190,7 @@ action.newTargetPowerSpawn = function(creep) {
             if (!powerSpawn) continue;
             var amount = 0;
             amount = powerSpawn.getNeeds(RESOURCE_ENERGY);
-            if (amount < 0) {
+            if (false&& amount < 0) {
```
Comment (Spedwards) [inline creep.action.reallocating.js]: The spacing on this is going to drive me crazy.

File: creep.action.reallocating.js [hunk 3]
```diff
@@ -217,7 +217,7 @@ action.newTargetPowerSpawn = function(creep) {
                 }
             }
             amount = powerSpawn.getNeeds(RESOURCE_POWER);
-            if (amount < 0) {
+            if (false&& amount < 0) {
```
Comment (Spedwards) [inline creep.action.reallocating.js]: See above.

File: creep.action.reallocating.js [hunk 4]
```diff
@@ -336,7 +336,7 @@ action.newTargetTerminal = function(creep) {
             if (amount > 0) {
                 // found a needed resource so check lower priority containers
                 if (DEBUG && TRACE) trace('Action', { actionName: 'reallocating', roomName: room.name, creepName: creep.name, structureId: terminal.id, resourceType: type, needs: amount });
-                if (room.storage && room.storage.store[type]) {
+                if (room.storage && room.storage.store[RESOURCE_ENERGY] && !(type==RESOURCE_ENERGY && room.storage.store[RESOURCE_ENERGY] > MIN_STORAGE_ENERGY[room.controller.level])) {
```
Comment (Spedwards) [inline creep.action.reallocating.js]: Please with the spacing. You trying to kill me?

File: room.js [hunk 5]
```diff
@@ -1070,14 +1124,46 @@ mod.extend = function(){
...
+            filter: (structure) => ( structure.structureType == STRUCTURE_POWER_SPAWN )
+        });
+
+        this.memory.powerSpawn = [];
+
+        // for each entry add to memory ( if not contained )
+        let add = (powerSpawn) => {
+            let powerSpawnData = this.memory.powerSpawn.find( (l) => l.id == powerSpawn.id );
+            if( !powerSpawnData ) {
+                this.memory.powerSpawn.push({
```
Comment (chobobobo) [inline room.js]: pushing an object to a list

File: room.js [hunk 6]
```diff
@@ -1070,14 +1124,46 @@ mod.extend = function(){
...
+            filter: (structure) => ( structure.structureType == STRUCTURE_NUKER )
+        });
+
+        this.memory.nuker = [];
+
+        // for each entry add to memory ( if not contained )
+        let add = (nuker) => {
+            let nukerData = this.memory.nuker.find( (l) => l.id == nuker.id );
+            if( !nukerData ) {
+                this.memory.nuker.push({
```
Comment (chobobobo) [inline room.js]: pushing to a list

**Comment by chobobobo** [comment]
Also, the room.memory.resources is undefined preventing my powerSpawn from processing power, but I don't know how/why it would be defined?

**Comment by Dasku** [comment]
powerSpawns will process without restriction, often leaving rooms with 10-100k energy in storage. Should we limit processing to when storage is over minimum energy?

**Comment by logxen** [comment]
that was already true, but having left it there they will take it lower to fill extensions ... it has since been raised to charge > 0.5 which is halfway between min and max.

**Comment by karlthepagan** [comment]
Having problems with debounce here: https://screeps.com/a/#!/room/E74S7

**Comment by karlthepagan** [comment]
removing isValidAction and isAddableAction from debounce fixed it

**Comment by logxen** [comment]
@karlthepagan ok I removed isvalidaction and isaddableaction

**Comment by chobobobo** [comment]
@logxen this is currently blocked waiting for nuker and powerSpawn to be properly defined in room.js, they are being pushed to memory as a list, but when they are loaded they are being considered ids.  Either we need to just push the id(my preferred solution as you can currently only ever have one per room) or we need to change the code that returns room.powerSpawn and room.nuker to properly read from the first element of the list if it exists.

**Comment by chobobobo** [comment]
Debounce is not working correctly in the case where the target is a controller container, the creeps are withdrawing and then transferring back in again until the controller container drops below a certain threshold.

**Comment by Dasku** [comment]
TypeError: Cannot read property '0' of undefined
TypeError: Cannot read property '0' of undefined
    at Room.fillARoomOrder (room:1516:47)
    at Room.terminalBroker (room:1625:32)
    at getEnvironment (room:2660:22)
    at /opt/engine/node_modules/lodash/index.js:3073:15
    at baseForOwn (/opt/engine/node_modules/lodash/index.js:2046:14)
    at /opt/engine/node_modules/lodash/index.js:3043:18
    at Function. (/opt/engine/node_modules/lodash/index.js:3346:13)
    at Function.m

**Comment by chobobobo** [comment]
I have a room with full storage, and creeps carrying power but not bringing it to the powerSpawn.

screen shot 2017-03-18 at 9 12 42 am

**Comment by chobobobo** [comment]
@karlthepagan I have made some changes here but I think this is generally working enough to be merged into /dev  there may be the odd bugfix here or there, but it's my opinion that it would be easier to do once the majority of this change has been merged.  I have resolved the issues I had with this branch.

**Comment by karlthepagan** [comment]
I agree this is needed. I've been missing it.

**Pull request merged**